### PR TITLE
Added rank of matrix in linear algebra

### PR DESCRIPTION
**Pull request by rudransh61**
Hi i am rudransh ,
I added a python file to find rank of matrix because i cannot find it anywhere in code base .
So, i thought that "lets contribute ".
Please try to pull this so that the set of algorithm will be completed.
If there is any error , inform me i will try to correct this.
AND AT LAST I WOULD LIKE TO SAY PLEASE SET IT UP  IN A VIRTUAL ENVIRONMENT SO THAT NEW VERSIONS OF REQUIREMENT AND PYTHON WILL NOT AFFECT ANYONE..
THANK YOU FOR READING THIS
A PROGRAMMER 
RUDRANSH BHARDWAJ

### CODE REVIEW

File: linear_algebra/src/Rank_of_Matrix.py
Comment (rohan472000) [inline linear_algebra/src/Rank_of_Matrix.py:5]: To resolve `ruff` errors, you should rename these functions to use lowercase letters. For example, you could rename "swapRows" to "swap_rows", "Row_Transformation" to "row_transformation", and "MatrixRank" to "matrix_rank".

**Review by rohan472000 — CHANGES_REQUESTED** [review]

**Review by rohan472000 — CHANGES_REQUESTED** [review]
The error message you received in `pre-commit` indicates that the file "Rank_of_Matrix.py" contains uppercase characters in its filename, which violates a common naming convention for Python files.

To resolve this error, you should rename the file to use lowercase letters instead. For example, you could rename it to "rank_of_matrix.py" instead.

File: linear_algebra/src/Rank_of_Matrix.py
Comment (rudransh61) [reply linear_algebra/src/Rank_of_Matrix.py:5]: Ok let me try

**Review by rohan472000 — CHANGES_REQUESTED** [review]
Kindly rename as I have suggested

File: linear_algebra/src/Rank_of_Matrix.py
Comment (rudransh61) [reply linear_algebra/src/Rank_of_Matrix.py:5]: Done👌

Comment (rohan472000) [reply linear_algebra/src/Rank_of_Matrix.py:5]: This is a bot, so no need to reply...just do as it suggests

**Comment by rudransh61** [comment]
I want 1 week because i had found some bugs in file 
so after some time it will ready for production

**Comment by rudransh61** [comment]
now all the errors and naming patterns are corrected except the name

**Review by rohan472000 — CHANGES_REQUESTED** [review]
it should be rank_of_mwtrix.py because it is a standard practice.

File: linear_algebra/src/Rank_of_Matrix.py
Comment (rudransh61) [reply linear_algebra/src/Rank_of_Matrix.py]: ok

**Comment by rudransh61** [comment]
finnally after trying for 2hrs i renamed the file

**Comment by rudransh61** [comment]
and thank you for your feedbacks

**Comment by rudransh61** [comment]
....is it ok to merge pull request or it is not okay

File: linear_algebra/src/rank_of_matrix.py
Comment (rohan472000) [inline linear_algebra/src/rank_of_matrix.py:3]: no need to give name.

File: linear_algebra/src/rank_of_matrix.py
Comment (rohan472000) [inline linear_algebra/src/rank_of_matrix.py:28]: add more unique doctests.

**Review by rohan472000 — CHANGES_REQUESTED** [review]

File: linear_algebra/src/rank_of_matrix.py
Comment (rudransh61) [reply linear_algebra/src/rank_of_matrix.py:28]: ok

Comment (rudransh61) [reply linear_algebra/src/rank_of_matrix.py:43]: removed my name and added more unique doctests

File: linear_algebra/src/rank_of_matrix.py
Comment (rohan472000) [inline linear_algebra/src/rank_of_matrix.py:43]: add doctest for two empty lists

**Review by rohan472000 — CHANGES_REQUESTED** [review]

File: linear_algebra/src/rank_of_matrix.py
Comment (rudransh61) [reply linear_algebra/src/rank_of_matrix.py:43]: oook
i have also added for [[1]]

**Comment by rudransh61** [comment]
.....can you merge this fork

**Comment by rohan472000** [comment]
I'm just a contributor, I don't have write access.

**Comment by rudransh61** [comment]
how can i ask him for review??

File: linear_algebra/src/rank_of_matrix.py
Comment (rudransh61) [reply linear_algebra/src/rank_of_matrix.py:47]: Thank you for your support sir

File: linear_algebra/src/rank_of_matrix.py
Comment (cclauss) [inline linear_algebra/src/rank_of_matrix.py:1]:
```suggestion
"""
Calculate the rank of a matrix.

See: https://en.wikipedia.org/wiki/Rank_(linear_algebra)
"""
```

File: linear_algebra/src/rank_of_matrix.py
Comment (cclauss) [inline linear_algebra/src/rank_of_matrix.py:10]: Please do not repeat the data types from the function signature. The problem with doing so is that the reader becomes confused if one changes and the other does not. Either describe what these variables are for or remove these comments.

```suggestion
        matrix: The matrix as a list of lists.
    Returns:
        The rank of the matrix.
```

File: linear_algebra/src/rank_of_matrix.py
Comment (cclauss) [inline linear_algebra/src/rank_of_matrix.py:4]:
```suggestion
def rank_of_matrix(matrix: list[list[int]]) -> int:
```

File: linear_algebra/src/rank_of_matrix.py
Comment (cclauss) [inline linear_algebra/src/rank_of_matrix.py:43]: Please add: `>>> rank_of_matrix([[]])`

File: linear_algebra/src/rank_of_matrix.py
Comment (rudransh61) [reply linear_algebra/src/rank_of_matrix.py:1]: ok sir

File: linear_algebra/src/rank_of_matrix.py
Comment (rudransh61) [reply linear_algebra/src/rank_of_matrix.py:47]: ok sir

Comment (rudransh61) [reply linear_algebra/src/rank_of_matrix.py:47]: Thanks for your suggestions

Comment (rudransh61) [reply linear_algebra/src/rank_of_matrix.py:47]: I hope this is ready to merge
and thanks to @rohan472000 and @cclauss for their feedbacks and suggestions

**Comment by rudransh61** [comment]
@cclauss please review and merge this (if it is ready)

**Comment by rudransh61** [comment]
I think now this is okk/?

**Comment by cclauss** [comment]
mypy.....................................................................Failed
- hook id: mypy
- exit code: 1

linear_algebra/src/rank_of_matrix.py:66: error: No overload variant of "__setitem__" of "list" matches argument types "int", "float"  [call-overload]
linear_algebra/src/rank_of_matrix.py:66: note: Possible overload variants:
linear_algebra/src/rank_of_matrix.py:66: note:     def __setitem__(self, SupportsIndex, int, /) -> None
linear_algebra/src/rank_of_matrix.py:66: note:     def __setitem__(self, slice, Iterable[int], /) -> None
Found 1 error in 1 file (checked 1186 source files)

**Comment by rudransh61** [comment]
i didnt understand ??? is there any error in code or something else???
I think there is some error while merging ???
or this is something strange ???
please tell me what is this , and what is hook id .....

I dont understand "mypy"

sorry sir for disturbing you but please help meeeeeee??

File: linear_algebra/src/rank_of_matrix.py
Comment (CaedenPH) [inline linear_algebra/src/rank_of_matrix.py:8]:
```suggestion
def rank_of_matrix(matrix: list[list[int | float]]) -> int:
```
This is required due to your multiplication returning a float

**Review by CaedenPH — COMMENTED** [review]
@rohan472000 Merge my commit and the checks will pass

**Review by rudransh61 — COMMENTED** [review]
nice
checks are showing error

**Comment by rudransh61** [comment]
oooooooooooo Thankakanakankankaaaasssss brooooo

FOR HELPING MEEEEEEEEEEEEEE
NOW 2 CHECKS ARE COMPLETED


you r genius ???? you are pro coderrr @CaedenPH 
THANKSSS

**Comment by rudransh61** [comment]
broooo you aree god

**Comment by rudransh61** [comment]
I think now it is ready to deploy??

**Comment by rudransh61** [comment]
Thanks to all reviews and other developers to help mee

**Review by cclauss — APPROVED** [review]
Let's do this!

**Pull request merged**
